comms/internal/rpcz/validator.go

```diff
@@ -256,3 +268,164 @@ func validateChatMembership(q db.Queryable, userId int32, chatId string) (db.Cha
 	})
 	return member, err
 }
+
+func validateNotBlocked(q db.Queryable, user1 int32, user2 int32) error {
+	blockedCount, err := queries.CountChatBlocks(q, context.Background(), queries.CountChatBlocksParams{
+		User1: user1,
+		User2: user2,
+	})
+	if err != nil {
+		return err
+	}
+	if blockedCount > 0 {
+		return errors.New("Cannot chat with a user you have blocked or user who has blocked you")
+	}
+
+	return nil
+}
+
+func calculateRateLimitCursor(timeframeVal string) (time.Time, error) {
+	var cursor time.Time
+	// Calculate cursor from rate limit timeframe
+	timeframe, err := strconv.Atoi(timeframeVal)
+	if err != nil {
+		return cursor, err
+	}
+	cursor = time.Now().UTC().Add(-time.Hour * time.Duration(timeframe))
+	return cursor, nil
+}
+
+func validateNewChatRateLimit(q db.Queryable, users []int32) error {
+	var err error
+
+	logger := config.Logger
+	timeframeVal := config.DefaultRateLimitRules[config.RateLimitTimeframeHours]
+	maxNumChatsVal := config.DefaultRateLimitRules[config.RateLimitMaxNumNewChats]
+
+	// Retrieve rate limit rules KV
+	var got nats.KeyValueEntry
+	jsc := jetstream.GetJetstreamContext()
+	kv, err := jsc.KeyValue(config.RateLimitRulesBucketName)
+	if err != nil {
+		goto KV_RETRIEVE_ERR
+	}
+
+	// Cursor for rate limit timeframe
+	got, err = kv.Get(config.RateLimitTimeframeHours)
+	if err != nil {
+		goto KV_RETRIEVE_ERR
+	}
+	timeframeVal = string(got.Value())
+
+	// Max num of new chats permitted per timeframe
+	got, err = kv.Get(config.RateLimitMaxNumNewChats)
+	if err != nil {
+		goto KV_RETRIEVE_ERR
+	}
+	maxNumChatsVal = string(got.Value())
+
+KV_RETRIEVE_ERR:
+	if err != nil {
+		logger.Warn("unable to retrive rate limit KV rules, using default values", "error", err)
+	}
+
+	cursor, err := calculateRateLimitCursor(timeframeVal)
+	maxNumChats, err := strconv.Atoi(maxNumChatsVal)
+	if err != nil {
+		return err
+	}
+
+	numChats, err := queries.MaxNumNewChatsSince(q, context.Background(), queries.MaxNumNewChatsSinceParams{
+		Users:  users,
+		Cursor: cursor,
+	})
+	if err != nil {
+		return err
+	}
+	if numChats >= maxNumChats {
+		return errors.New("An invited user has exceeded the maximum number of new chats")
+	}
+
+	return nil
+}
+
+func validateNewMessageRateLimit(q db.Queryable, userId int32, chatId string) error {
+	var err error
+
+	logger := config.Logger
+	timeframeVal := config.DefaultRateLimitRules[config.RateLimitTimeframeHours]
+	maxNumMessagesVal := config.DefaultRateLimitRules[config.RateLimitMaxNumMessages]
+	maxNumMessagesPerRecipientVal := config.DefaultRateLimitRules[config.RateLimitMaxNumMessagesPerRecipient]
+
+	// Retrieve rate limit rules KV
+	var got nats.KeyValueEntry
+	jsc := jetstream.GetJetstreamContext()
+	kv, err := jsc.KeyValue(config.RateLimitRulesBucketName)
+	if err != nil {
+		goto KV_RETRIEVE_ERR
+	}
+
+	// Cursor for rate limit timeframe
+	got, err = kv.Get(config.RateLimitTimeframeHours)
```

Comment:
I'd write a helper function to encapsulate getting value, casting, or returning default if err.

Something like: `getRateLimit(kv nats.KV, name string, fallback int) int`

Also should be int or float return instead of string, since we just call `Atoi` later, right?

Reply:
https://github.com/AudiusProject/audius-protocol/pull/4542/commits/05ad5c3883cd452d8759a1ff8fd53aaa7323a5b8